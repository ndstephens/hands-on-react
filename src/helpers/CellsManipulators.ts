import { Cell, CellState, Coords, Field } from './Field';

// Get all coordinates of a Cell's 8 surrounding neighbors
export const getNeighborItems = ([y, x]: Coords): Record<string, Coords> => {
  return {
    top: [y - 1, x],
    topRight: [y - 1, x + 1],
    right: [y, x + 1],
    bottomRight: [y + 1, x + 1],
    bottom: [y + 1, x],
    bottomLeft: [y + 1, x - 1],
    left: [y, x - 1],
    topLeft: [y - 1, x - 1],
  };
};

// Check that a Cell is actually in the field (aka on the board)
export const checkItemInField = ([y, x]: Coords, field: Field): boolean => {
  if (y < 0 || y >= field.length || x < 0 || x >= field.length) {
    return false;
  }
  return true;
};

// Increment all of a Cell's (non-bomb) neighbor's values by 1
// Use to indicate number of bombs surrounding a Cell
export const incrementNeighbors = (coords: Coords, field: Field): Field => {
  const neighbors = getNeighborItems(coords);

  Object.values(neighbors).forEach(([y, x]) => {
    // verify it's within the field coords and also not a bomb (nor can become a bomb)
    if (checkItemInField([y, x], field) && field[y]![x]! < CellState.bomb - 1) {
      field[y]![x] = (field[y]![x]! + 1) as Cell;
    }
  });

  return field;
};
