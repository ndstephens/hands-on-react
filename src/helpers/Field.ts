import { incrementNeighbors } from './CellsManipulators';

export const PROBABILITY_ERROR_MSG = 'Probability must be between 0 and 1';

export type Cell = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12;
export type Field = Cell[][];
export type Coords = [number, number]; // [y, x]

type CellStateNames = 'empty' | 'bomb' | 'hidden' | 'flag' | 'weakFlag';
type CellState = {
  [key in CellStateNames]: Cell;
};
export const CellState: CellState = {
  empty: 0,
  bomb: 9,
  hidden: 10,
  flag: 11,
  weakFlag: 12,
};

// Create a field with empty cells
export const emptyFieldGenerator = (
  size: number,
  state: Cell = CellState.empty
): Field => {
  return new Array(size).fill(null).map(() => new Array(size).fill(state));
};

// Create a field with a bomb probability (between 0 and 1)
export const fieldGenerator = (size: number, probability: number): Field => {
  if (probability < 0 || probability > 1) {
    throw new Error(PROBABILITY_ERROR_MSG);
  }

  let unprocessedCells = size * size;
  let cellsWithBombs = unprocessedCells * probability;
  const field = emptyFieldGenerator(size);

  if (cellsWithBombs === 0) return field;

  for (let y = 0; y < size; y++) {
    for (let x = 0; x < size; x++) {
      if (cellsWithBombs / unprocessedCells > Math.random()) {
        field[y]![x] = CellState.bomb;
        incrementNeighbors([y, x], field);
        cellsWithBombs--;
      }
      unprocessedCells--;
    }
  }

  return field;
};
