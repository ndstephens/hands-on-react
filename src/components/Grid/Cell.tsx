import styled from '@emotion/styled';

import { Cell as CellType, CellState, Coords } from '@/helpers/Field';
import { useMouseDown } from '@/hooks/useMouseDown';

export interface CellProps {
  /**
   * Cell status...number between 0 and 12
   */
  children: CellType;
  /**
   * Cell coordinates
   */
  coords: Coords;
  /**
   * onClick Cell handler
   */
  onClick: (coords: Coords) => void;
  /**
   * onContextMenu Cell handler
   */
  onContextMenu: (coords: Coords) => void;
}

export const Cell = ({ children, coords, ...restProps }: CellProps) => {
  const [mouseDown, onMouseDown, onMouseUp] = useMouseDown();

  const onClick = () => restProps.onClick(coords);

  const onContextMenu = (e: React.MouseEvent<HTMLDivElement>) => {
    e.preventDefault();
    if (checkCellIsActive(children)) {
      restProps.onContextMenu(coords);
    }
  };

  const props = {
    onClick,
    onContextMenu,
    onMouseDown,
    onMouseUp,
    onMouseLeave: onMouseUp,
    mouseDown,
    'data-testid': `${children}_${coords}`,
  };

  return <CellComponent {...props}>{children}</CellComponent>;
};

/* =============================================
              UTILITY FUNCTIONS
============================================= */
export const checkCellIsActive = (cell: CellType): boolean => {
  return [CellState.hidden, CellState.flag, CellState.weakFlag].includes(cell);
};

/* =============================================
                CELL COMPONENT
============================================= */
interface CellComponentProps {
  children: CellType;
  onClick: () => void;
  onContextMenu: (e: React.MouseEvent<HTMLDivElement>) => void;
  onMouseDown: () => void;
  onMouseUp: () => void;
  onMouseLeave: () => void;
  mouseDown: boolean;
  'data-testid'?: string;
}
const CellComponent = ({ children, ...restProps }: CellComponentProps) => {
  const nonActiveCellProps = {
    onContextMenu: restProps.onContextMenu,
    'data-testid': restProps['data-testid'],
  };

  switch (children) {
    case CellState.empty:
      return <EmptyFrame {...nonActiveCellProps} />;
    case CellState.bomb:
      return (
        <BombFrame {...nonActiveCellProps}>
          <Bomb />
        </BombFrame>
      );
    case CellState.hidden:
      return <ClosedFrame {...restProps} />;
    case CellState.flag:
      return (
        <ClosedFrame {...restProps}>
          <Flag />
        </ClosedFrame>
      );
    case CellState.weakFlag:
      return (
        <ClosedFrame {...restProps}>
          <WeakFlag />
        </ClosedFrame>
      );
    default:
      return <RevealedFrame {...nonActiveCellProps}>{children}</RevealedFrame>;
  }
};

/* =============================================
            STYLED-COMPONENTS
============================================= */
interface ClosedFrameProps {
  mouseDown?: boolean;
}

const ClosedFrame = styled.div<ClosedFrameProps>`
  display: flex;
  align-items: center;
  justify-content: center;
  user-select: none;
  cursor: pointer;
  width: 1.8vw;
  height: 1.8vw;
  background-color: #d1d1d1;
  border: 0.6vh solid transparent;
  border-color: ${({ mouseDown = false }) =>
    mouseDown ? 'transparent' : ' white #9e9e9e #9e9e9e white'};

  &:hover {
    filter: brightness(1.1);
  }
`;

const EmptyFrame = styled(ClosedFrame)`
  border-color: #ddd;
  cursor: default;

  &:hover {
    filter: brightness(1);
  }
`;

const BombFrame = styled(EmptyFrame)`
  background-color: #ec433c;
`;

const Bomb = styled.div`
  border-radius: 50%;
  width: 60%;
  height: 60%;
  background-color: #333;
`;

const Flag = styled.div`
  width: 0;
  height: 0;
  border-top: 0.5vw solid transparent;
  border-bottom: 0.5vw solid transparent;
  border-left: 0.5vw solid #ec433c;
`;

const WeakFlag = styled(Flag)`
  border-left-color: #f19996;
`;

const transparent = 'rgba(0,0,0,0)';
const colors: { [key in CellType]: string } = {
  0: transparent,
  1: '#2a48ec',
  2: '#2bb13d',
  3: '#ec6561',
  4: '#233db7',
  5: '#a6070f',
  6: '#e400af',
  7: '#906a02',
  8: '#fa0707',
  9: transparent,
  10: transparent,
  11: transparent,
  12: transparent,
};

const RevealedFrame = styled(EmptyFrame)`
  color: ${({ children }) => colors[children as CellType] ?? transparent};
  font-size: 1.2vw;
`;
