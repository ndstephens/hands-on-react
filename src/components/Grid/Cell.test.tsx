import { createEvent, fireEvent, render, screen } from '@testing-library/react';

import { CellState, Coords } from '@/helpers/Field';

import { Cell, CellProps, checkCellIsActive, ClosedFrame } from './Cell';

describe('Cell component', () => {
  // Default props for all tests
  const coords: Coords = [1, 1];

  const props: Omit<CellProps, 'children'> = {
    coords,
    onClick: jest.fn(),
    onContextMenu: jest.fn(),
  };

  // Loop through all Cell types and run tests on each
  for (let cell = CellState.empty; cell <= CellState.weakFlag; cell++) {
    it(`Cell renders correctly - Cell: ${cell}`, () => {
      const { asFragment } = render(<Cell {...props}>{cell}</Cell>);

      expect(asFragment()).toMatchSnapshot();
    });

    it('Closed Frame styled-component renders correctly', () => {
      const { asFragment } = render(<ClosedFrame mouseDown={true} />);

      expect(asFragment()).toMatchSnapshot();
    });

    it(`Check context menu preventDefault works for every cell type - Cell: ${cell}`, () => {
      render(<Cell {...props}>{cell}</Cell>);

      const cellComponent = screen.getByTestId(`${cell}_${coords}`);

      // const isDefaultPrevented = fireEvent.contextMenu(cellComponent);
      // expect(isDefaultPrevented).toBe(false);

      const contextMenuEvent = createEvent.contextMenu(cellComponent);
      fireEvent(cellComponent, contextMenuEvent);
      expect(contextMenuEvent.defaultPrevented).toBe(true);
    });

    it(`onClick and onContextMenu handlers should only be called for active cells - Cell: ${cell}`, () => {
      render(<Cell {...props}>{cell}</Cell>);

      const cellComponent = screen.getByTestId(`${cell}_${coords}`);

      fireEvent.click(cellComponent);
      fireEvent.contextMenu(cellComponent);

      if (checkCellIsActive(cell)) {
        expect(props.onClick).toHaveBeenCalledTimes(1);
        expect(props.onContextMenu).toHaveBeenCalledTimes(1);
      } else {
        expect(props.onClick).toHaveBeenCalledTimes(0);
        expect(props.onContextMenu).toHaveBeenCalledTimes(0);
      }
    });
  }
});
