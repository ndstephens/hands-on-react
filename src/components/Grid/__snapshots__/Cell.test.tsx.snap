// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell component Cell renders correctly - Cell: 0 1`] = `
<DocumentFragment>
  <div
    class="css-1ux1iji"
    data-testid="0_1,1"
  />
</DocumentFragment>
`;

exports[`Cell component Cell renders correctly - Cell: 1 1`] = `
<DocumentFragment>
  <div
    class="css-1nnrvqf"
    data-testid="1_1,1"
  >
    1
  </div>
</DocumentFragment>
`;

exports[`Cell component Cell renders correctly - Cell: 2 1`] = `
<DocumentFragment>
  <div
    class="css-53i6ep"
    data-testid="2_1,1"
  >
    2
  </div>
</DocumentFragment>
`;

exports[`Cell component Cell renders correctly - Cell: 3 1`] = `
<DocumentFragment>
  <div
    class="css-160dk3x"
    data-testid="3_1,1"
  >
    3
  </div>
</DocumentFragment>
`;

exports[`Cell component Cell renders correctly - Cell: 4 1`] = `
<DocumentFragment>
  <div
    class="css-fhs7xs"
    data-testid="4_1,1"
  >
    4
  </div>
</DocumentFragment>
`;

exports[`Cell component Cell renders correctly - Cell: 5 1`] = `
<DocumentFragment>
  <div
    class="css-xb2jm"
    data-testid="5_1,1"
  >
    5
  </div>
</DocumentFragment>
`;

exports[`Cell component Cell renders correctly - Cell: 6 1`] = `
<DocumentFragment>
  <div
    class="css-8xfpdp"
    data-testid="6_1,1"
  >
    6
  </div>
</DocumentFragment>
`;

exports[`Cell component Cell renders correctly - Cell: 7 1`] = `
<DocumentFragment>
  <div
    class="css-130cupr"
    data-testid="7_1,1"
  >
    7
  </div>
</DocumentFragment>
`;

exports[`Cell component Cell renders correctly - Cell: 8 1`] = `
<DocumentFragment>
  <div
    class="css-fq6n7j"
    data-testid="8_1,1"
  >
    8
  </div>
</DocumentFragment>
`;

exports[`Cell component Cell renders correctly - Cell: 9 1`] = `
<DocumentFragment>
  <div
    class="css-4r8q4z"
    data-testid="9_1,1"
  >
    <div
      class="css-2eux8l"
    />
  </div>
</DocumentFragment>
`;

exports[`Cell component Cell renders correctly - Cell: 10 1`] = `
<DocumentFragment>
  <div
    class="css-1iw1ast"
    data-testid="10_1,1"
  />
</DocumentFragment>
`;

exports[`Cell component Cell renders correctly - Cell: 11 1`] = `
<DocumentFragment>
  <div
    class="css-1iw1ast"
    data-testid="11_1,1"
  >
    <div
      class="css-oefih3"
    />
  </div>
</DocumentFragment>
`;

exports[`Cell component Cell renders correctly - Cell: 12 1`] = `
<DocumentFragment>
  <div
    class="css-1iw1ast"
    data-testid="12_1,1"
  >
    <div
      class="css-13jo0mv"
    />
  </div>
</DocumentFragment>
`;

exports[`Cell component Closed Frame styled-component renders correctly 1`] = `
<DocumentFragment>
  <div
    class="css-1mjcmhi"
  />
</DocumentFragment>
`;

exports[`Cell component Closed Frame styled-component renders correctly 2`] = `
<DocumentFragment>
  <div
    class="css-1mjcmhi"
  />
</DocumentFragment>
`;

exports[`Cell component Closed Frame styled-component renders correctly 3`] = `
<DocumentFragment>
  <div
    class="css-1mjcmhi"
  />
</DocumentFragment>
`;

exports[`Cell component Closed Frame styled-component renders correctly 4`] = `
<DocumentFragment>
  <div
    class="css-1mjcmhi"
  />
</DocumentFragment>
`;

exports[`Cell component Closed Frame styled-component renders correctly 5`] = `
<DocumentFragment>
  <div
    class="css-1mjcmhi"
  />
</DocumentFragment>
`;

exports[`Cell component Closed Frame styled-component renders correctly 6`] = `
<DocumentFragment>
  <div
    class="css-1mjcmhi"
  />
</DocumentFragment>
`;

exports[`Cell component Closed Frame styled-component renders correctly 7`] = `
<DocumentFragment>
  <div
    class="css-1mjcmhi"
  />
</DocumentFragment>
`;

exports[`Cell component Closed Frame styled-component renders correctly 8`] = `
<DocumentFragment>
  <div
    class="css-1mjcmhi"
  />
</DocumentFragment>
`;

exports[`Cell component Closed Frame styled-component renders correctly 9`] = `
<DocumentFragment>
  <div
    class="css-1mjcmhi"
  />
</DocumentFragment>
`;

exports[`Cell component Closed Frame styled-component renders correctly 10`] = `
<DocumentFragment>
  <div
    class="css-1mjcmhi"
  />
</DocumentFragment>
`;

exports[`Cell component Closed Frame styled-component renders correctly 11`] = `
<DocumentFragment>
  <div
    class="css-1mjcmhi"
  />
</DocumentFragment>
`;

exports[`Cell component Closed Frame styled-component renders correctly 12`] = `
<DocumentFragment>
  <div
    class="css-1mjcmhi"
  />
</DocumentFragment>
`;

exports[`Cell component Closed Frame styled-component renders correctly 13`] = `
<DocumentFragment>
  <div
    class="css-1mjcmhi"
  />
</DocumentFragment>
`;
